packages/pipeline-editor/src/validation.spec.ts

```diff
@@ -0,0 +1,343 @@
+  //                                                          ▲          │
+  //                                                          ╰────l9────╯
+  {
+    it: 'should handle long forks',
+    given: [
+      { id: 'l0', srcNodeId: 'a', trgNodeId: 'b', type: 'nodeLink' },
+      { id: 'l1', srcNodeId: 'b', trgNodeId: 'c', type: 'nodeLink' },
+      { id: 'l2', srcNodeId: 'c', trgNodeId: 'd', type: 'nodeLink' },
+      { id: 'l3', srcNodeId: 'd', trgNodeId: 'e', type: 'nodeLink' },
+      { id: 'l4', srcNodeId: 'c', trgNodeId: 'f', type: 'nodeLink' },
+      { id: 'l5', srcNodeId: 'f', trgNodeId: 'g', type: 'nodeLink' },
+      { id: 'l6', srcNodeId: 'g', trgNodeId: 'h', type: 'nodeLink' },
+      { id: 'l7', srcNodeId: 'h', trgNodeId: 'i', type: 'nodeLink' },
+      { id: 'l8', srcNodeId: 'i', trgNodeId: 'j', type: 'nodeLink' },
+      { id: 'l9', srcNodeId: 'j', trgNodeId: 'i', type: 'nodeLink' }
+    ],
+    expected: ['l8', 'l9']
+  },
+
+  // ╭───╮      ╭───╮      ╭───╮      ╭───╮      ╭───╮
+  // │ a │──l0─►│ b │──l1─►│ c │──l8─►│ i │──l9─►│ j │
+  // ╰───╯      ╰───╯      ╰───╯      ╰───╯      ╰───╯
+  //                         │
+  //                        l2
+  //                         ▼
+  //                       ╭───╮      ╭───╮      ╭───╮      ╭───╮      ╭───╮
+  //                       │ d │──l3─►│ e │──l4─►│ f │──l5─►│ g │──l6─►│ h │
+  //                       ╰───╯      ╰───╯      ╰───╯      ╰───╯      ╰───╯
+  //                                                          ▲          │
+  //                                                          ╰────l7────╯
+  {
+    it: 'should handle long forks - shuffled',
+    given: [
+      { id: 'l0', srcNodeId: 'a', trgNodeId: 'b', type: 'nodeLink' },
+      { id: 'l1', srcNodeId: 'b', trgNodeId: 'c', type: 'nodeLink' },
+      { id: 'l2', srcNodeId: 'c', trgNodeId: 'd', type: 'nodeLink' },
+      { id: 'l3', srcNodeId: 'd', trgNodeId: 'e', type: 'nodeLink' },
+      { id: 'l4', srcNodeId: 'e', trgNodeId: 'f', type: 'nodeLink' },
+      { id: 'l5', srcNodeId: 'f', trgNodeId: 'g', type: 'nodeLink' },
+      { id: 'l6', srcNodeId: 'g', trgNodeId: 'h', type: 'nodeLink' },
+      { id: 'l7', srcNodeId: 'h', trgNodeId: 'g', type: 'nodeLink' },
+      { id: 'l8', srcNodeId: 'c', trgNodeId: 'i', type: 'nodeLink' },
+      { id: 'l9', srcNodeId: 'i', trgNodeId: 'j', type: 'nodeLink' }
+    ],
+    expected: ['l6', 'l7']
+  },
+
+  // ╭───╮      ╭───╮      ╭───╮      ╭───╮      ╭───╮
+  // │ a │──l0─►│ b │──l1─►│ c │──l5─►│ f │──l6─►│ g │
+  // ╰───╯      ╰───╯      ╰───╯      ╰───╯      ╰───╯
+  //                         │          ▲
+  //                        l2          l4
+  //                         ▼          │
+  //                       ╭───╮      ╭───╮
+  //                       │ d │──l3─►│ e │
+  //                       ╰───╯      ╰───╯
+  {
+    it: 'should not have a bug that a previous implementation had',
```

Comment:
what was the bug? That might be a better description for someone seeing this for the first time...

Reply:
I actually don't know haha... I just had this graph sketched out on a piece of paper of other graphs that had issues and couldn't remember what the issue was when writing tests so I gave it a vague name 😬

It either incorrectly indicated a cycle or had caused an infinite loop, but also could have not been and bug and wrote it down for another reason 🤷‍♂️